### PR TITLE
BUG: fixes indexing with monotonic decreasing DTI (#19362)

### LINKED ISSUE #19362: BUG: indexing with monotonic decreasing DTI
Hi there, I'm trying to add a column of time using pandas. But some dates doesn't work fine in string format. I'm working on jupyter notebook with Anaconda 4.3.0

Here's what I'm trying to do:

```python
data1 = data.head()
date_list = ["2018-01-02", "2017-02-10","2016-03-10", "2015-03-15", "2014-03-16"]
date_index = pd.to_datetime(date_list)
data1['date'] = date_index
data1
```
And the output is shown below.
image

Then I try to set 'date' as the index of this data frame.
```python
data1 = data1.set_index("date")
print(data1.index)
data1
```
This looks reasonable, too.
image

Something weird happens when I try to index data using:
```python
data1.loc['2018-01-02']
```
It raises KeyError!!! 
image

But other dates work fine. For example:
```python
data1.loc['2015-03-15']
```
image

Why is that? 

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.13.3
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.2.2
numexpr: 2.6.1
matplotlib: 2.0.0
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.2
bs4: 4.5.3
html5lib: 0.9999999
httplib2: None
apiclient: None
sqlalchemy: 1.1.5
pymysql: None
psycopg2: None
jinja2: 2.9.4
boto: 2.45.0
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by mapehe**
- [x] closes #19362
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Seems that #19362 is caused by the fact that `np.any([0])` is false. Here's a suggested solution. Also added a test.

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -342,7 +342,7 @@ def _format_with_header(self, header, **kwargs):
     def __contains__(self, key):
         try:
             res = self.get_loc(key)
-            return is_scalar(res) or type(res) == slice or np.any(res)
+            return is_scalar(res) or type(res) == slice or np.any(res + 1)
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: can you change ``type(res) == slice`` to ``isinstance(res, slice)``.

I think condition could be (for the last clause)
``is_list_like(res) and len(res)``. just need to account for an indexer, not the character of it

File: pandas/tests/indexing/test_datetime.py [hunk 2]
```diff
@@ -252,3 +252,11 @@ def test_series_partial_set_period(self):
...
             result = ser.loc[keys]
         tm.assert_series_equal(result, exp)
+
+    def test_monotone_DTI_indexing_bug(self):
+        # GH 19362
+
+        df = pd.DataFrame({'A': [1, 2, 3]},
+                          index=pd.date_range('20170101',
+                                              periods=3)[::-1])
+        df.loc['2017-01-03']
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: can you assert on the result of this

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note (indexing bug fix section)

File: pandas/tests/indexing/test_datetime.py [hunk 3]
```diff
@@ -252,3 +252,11 @@ def test_series_partial_set_period(self):
                                         check_stacklevel=False):
             result = ser.loc[keys]
         tm.assert_series_equal(result, exp)
+
+    def test_monotone_DTI_indexing_bug(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: this test belongs in one of:

```
pandas/tests/indexes/datetimes/test_datetime.py
pandas/tests/indexes/datetimes/test_partial_slicing.py
```

**Comment by mapehe** [comment]
Thanks for the helpful comments @jreback, I think the updated commit should address the points.

File: pandas/tests/indexes/datetimes/test_datetime.py [hunk 4]
```diff
@@ -368,3 +368,14 @@ def test_factorize_dst(self):
     def test_unique(self, arr, expected):
         result = arr.unique()
         tm.assert_index_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_datetime.py]: can you move near the other tests?

also can you add all of the examples from the original issue (e.g. the working ones). so this is a complete example.

can you add a comment on what is being tested here.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some small comments. ping on green.

File: pandas/core/indexes/datetimelike.py [hunk 5]
```diff
@@ -342,7 +342,8 @@ def _format_with_header(self, header, **kwargs):
     def __contains__(self, key):
         try:
             res = self.get_loc(key)
-            return is_scalar(res) or type(res) == slice or np.any(res)
+            return is_scalar(res) or isinstance(res, slice) or \
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimelike.py]: Typically we use parenthesis to break long lines, not `\`.

**Review by TomAugspurger — COMMENTED** [review]
Small style comment. @mapehe can you update fixing that & @jreback's suggestions (and fix the merge conflict). Doing a release this week.

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (mapehe) [reply pandas/core/indexes/datetimelike.py]: Changed that

File: pandas/tests/indexes/datetimes/test_datetime.py (hunk 4 above)
Comment (mapehe) [reply pandas/tests/indexes/datetimes/test_datetime.py]: > can you move near the other tests?

Sorry, didn't quite get this one, but tried to move this closer to other tests that seem a little similar.

> also can you add all of the examples from the original issue (e.g. the working ones). so this is a complete example.

There wasn't a lot of stuff in the original issue, but I added what the person who posted that issue was attempting to do. 

> can you add a comment on what is being tested here.

Added a short comment.

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (mapehe) [reply pandas/core/indexes/datetimelike.py]: Also fixed the conflict

**Comment by mapehe** [comment]
https://github.com/pandas-dev/pandas/pull/20677/commits/3c4f38b005c7d7a4e2e66d180a536acc2b9c57a3 was all green @jreback

**Comment by jreback** [comment]
thanks @mapehe!

**Comment by mapehe** [comment]
Np!

**Pull request merged**